client/src/unifyfs-fixed.c

```diff
@@ -471,6 +473,82 @@ static int unifyfs_split_index(
     return UNIFYFS_SUCCESS;
 }
 
+/* Add the metadata for a single write to the index */
+static int unifyfs_logio_add_write_meta_to_index(int gfid, off_t file_pos,
+    off_t log_pos, size_t length)
+{
+    /* define an new index entry for this write operation */
+    unifyfs_index_t cur_idx;
+    cur_idx.gfid     = gfid;
+    cur_idx.file_pos = file_pos;
+    cur_idx.log_pos  = log_pos;
+    cur_idx.length   = length;
+
+    /* lookup number of existing index entries */
+    off_t num_entries = *(unifyfs_indices.ptr_num_entries);
+
+    /* get pointer to index array */
+    unifyfs_index_t* idxs = unifyfs_indices.index_entry;
+
+    /* attempt to coalesce contiguous index entries if we
+     * have an existing index in the buffer */
+    if (num_entries > 0) {
+        /* get pointer to last element in index array */
+        unifyfs_index_t* prev_idx = &idxs[num_entries - 1];
+
+        /* attempt to coalesce current index with last index,
+         * updates fields in last index and current index
+         * accordingly */
+        unifyfs_coalesce_index(prev_idx, &cur_idx,
+            unifyfs_key_slice_range);
+    }
+
+    /* add new index entries if needed */
+    while (cur_idx.length > 0) {
+        /* remaining entries we can fit in the shared memory region */
+        off_t remaining_entries = unifyfs_max_index_entries - num_entries;
+
+        /* if we have filled the key/value buffer, flush it to server */
+        if (0 == remaining_entries) {
+            /* index buffer is full, flush it */
+            int ret = invoke_client_fsync_rpc(cur_idx.gfid);
+            if (ret != UNIFYFS_SUCCESS) {
+                /* something went wrong when trying to flush key/values */
+                LOGERR("failed to flush key/value index to server");
+                return UNIFYFS_ERROR_IO;
+            }
+
+            /* flushed, clear buffer and refresh number of entries
+             * and number remaining */
+            num_entries = 0;
+            *(unifyfs_indices.ptr_num_entries) = num_entries;
+            remaining_entries = unifyfs_max_index_entries - num_entries;
```

Comment:
Let's we rewrite this code to call your new clear_index function now that we have it